futures-util/src/io/split.rs

```diff
@@ -21,7 +22,9 @@ fn lock_and_then<T, U, E, F>(lock: &BiLock<T>, cx: &mut task::Context, f: F) ->
     where F: FnOnce(&mut T, &mut task::Context) -> Poll<Result<U, E>>
 {
     match lock.poll_lock(cx) {
-        Poll::Ready(ref mut l) => f(l, cx),
+        // Safety: the value behind the bilock used by `ReadHalf` and `WriteHalf` is never exposed
+        // as a `PinMut` anywhere other than here as a way to get to `&mut`.
```

Comment:
This is pretty sketchy and I can imagine it might not hold up in a types-as-contracts-based model that formally understood `PinMut` as a new typestate, but it seems "fine". The alternatives are (a) take `PinMut` in the `AsyncRead` and `AsyncWrite` methods, (b) put an `Unpin` bound on the `AsyncRead` and `AsyncWrite` types used in a split, (c) make two versions of `BiLock`, one which pins its innards, and one which does not.